cupy/core/raw.pyx

```diff
@@ -179,3 +187,68 @@ def _get_raw_kernel(code, name, options=()):
     module = cupy.core.core.compile_with_cache(code, options,
                                                prepend_cupy_headers=False)
     return module.get_function(name)
+
+
+cdef class RawModule:
+    """User-defined custom module.
+
+    This class can be used to either compile raw CUDA sources or load CUDA
+    modules (\*.cubin). This class is useful when a number of CUDA kernels in
+    the same source need to be retrieved.
+
+    For the former case, the CUDA source code is compiled when initializing a
+    new instance of this class, and the kernels can be retrieved by calling
+    :meth:`get_function`, which will return an instance of :class:`RawKernel`.
+    (Same as in :class:`RawKernel`, the generated binary is also cached.)
+
+    For the latter case, an existing CUDA binary (\*.cubin) can be loaded by
+    providing its path, and kernels therein can be retrieved similarly.
+
+    Args:
+        code_or_path (str): CUDA source code or path to cubin.
```

Comment:
Just some comment from @asi1024 ,
Is this safe?, If the path is set to be soemthing like `*.cubin`  if the compilation fails, some of the errors can be hard to understand.
what do you think?

Reply:
If the path is set to a cubin, no compilation would happen. It's the `cupy.cuda.driver.moduleLoad()` that gets invoked, and a `CUDADriverError` is raised (`CUDA_ERROR_FILE_NOT_FOUND`):
```python
In [1]: import cupy as cp

In [2]: mod = cp.RawModule("/hoejfwj/cwfoq.cubin")
---------------------------------------------------------------------------
CUDADriverError                           Traceback (most recent call last)
<ipython-input-2-2514fdf9b024> in <module>()
----> 1 mod = cp.RawModule("/hoejfwj/cwfoq.cubin")

/GPFS/XF03ID1/home/leofang/test/cupy2/cupy/core/raw.pyx in cupy.core.raw.RawModule.__init__()

/GPFS/XF03ID1/home/leofang/test/cupy2/cupy/cuda/function.pyx in cupy.cuda.function.Module.load_file()

/GPFS/XF03ID1/home/leofang/test/cupy2/cupy/cuda/function.pyx in cupy.cuda.function.Module.load_file()

/GPFS/XF03ID1/home/leofang/test/cupy2/cupy/cuda/driver.pyx in cupy.cuda.driver.moduleLoad()

/GPFS/XF03ID1/home/leofang/test/cupy2/cupy/cuda/driver.pyx in cupy.cuda.driver.check_status()

CUDADriverError: CUDA_ERROR_FILE_NOT_FOUND: file not found
```
But this reminded me that I could perhaps do some error handling. Let me try some ideas.

Reply:
I added two tests for possible errors, the error messages of which are self-explanatory I think, so I'd say just let the error propagate instead of adding more guards. 

To summarize, I can think of 4 possibilities that could go wrong in `RawModule`:
1. Invalid CUDA code: A `cupy.cuda.compiler.CompileException` will be raised;
2. Invalid compiler options: same exception as above;
3. Invalid cubin path: A `cupy.cuda.driver.CUDADriverError` will be raised;
4. Invalid kernel name: same exception as above

all of which are now tested in aa5199e. Could you come up with any other possibilities? (It's mid night now, so I am very likely not thinking clearly...)

Thanks!

Reply:
Thanks for the super detailed explanation.

What would happen when using wildcards on the path to feed multiple files at once?